src/LoanManager.ts

```diff
@@ -50,12 +71,134 @@ export class LoanManager extends AbstractContract {
+        const gasEstimate: number = COLLECT_BASE_GAS + COLLECT_ONE_GAS * loansToCollect.length;
+
+        const loanIdsToCollect: number[] = loansToCollect.map((loan:Loan) => loan.id);
+
+        const web3Tx: TransactionObject<void> = this.instance.methods.collect(loanIdsToCollect);
+
+        return new Transaction(this.ethereumConnection, web3Tx, {
+            gasLimit: gasEstimate,
+            from: userAccount
+        });
+    }
+
+    public async getAllLoans(): Promise<Loan[]> {
+        try {
+            const loanManagerInstance:LoanManagerInstance = this.instance;
+            const chunkSize:number = isLoanManagerV0(loanManagerInstance) ? LEGACY_CONTRACTS_CHUNK_SIZE : CHUNK_SIZE;
+
+            const loanCount: number = await this.getLoanCount();
+            const tokenAddress: string = await this.tokenAddress;
+            let loansToCollect: Loan[] = [];
+
+            const queryCount: number = Math.ceil(loanCount / chunkSize);
+            for (let i: number = 0; i < queryCount; i++) {
+                const loansArray: string[][] = isLoanManagerV0(loanManagerInstance)
+                    ? await loanManagerInstance.methods.getLoans(i * chunkSize).call()
+                    : await loanManagerInstance.methods.getLoans(i * chunkSize, chunkSize).call();
+                loansToCollect = loansToCollect.concat(
+                    loansArray.map((loan: ILoanTuple) => new Loan(loan, this.address, tokenAddress))
+                );
+            }
+
+            return loansToCollect;
+        } catch (error) {
+            throw new Error("fetchAllLoansTx failed.\n" + error);
+        }
+    }
+
+    public getLoanCount(): Promise<number> {
+        return this.instance.methods
+            .getLoanCount()
+            .call()
+            .then((res: string) => parseInt(res, 10));
+    }
+
+    public addExtraCollateral(loan: Loan, weiAmount: Wei, userAccount: string): Transaction {
+        if (loan.isMarginLoan && !isLoanManagerPreMarginLoan(this.instance)) {
+            const web3Tx: TransactionObject<void> = this.instance.methods.addExtraCollateral(loan.id);
+            return new Transaction(this.ethereumConnection, web3Tx, {
+                from: userAccount,
+                value: weiAmount.amount
+            });
+        } else {
+            throw new AugmintJsError('invalid call to addExtraCollateral');
+        }
+    }
 
     private async getProducts(onlyActive: boolean): Promise<LoanProduct[]> {
-        // @ts-ignore  TODO: how to detect better without ts-ignore?
-        const isLegacyLoanContractWithChunkSize: boolean = typeof this.instance.methods.CHUNK_SIZE === "function";
-        const chunkSize: number = isLegacyLoanContractWithChunkSize ? LEGACY_CONTRACTS_CHUNK_SIZE : CHUNK_SIZE;
+        const chunkSize: number = isLoanManagerV0(this.instance) ? LEGACY_CONTRACTS_CHUNK_SIZE : CHUNK_SIZE;
 
         const productCount: number = await this.instance.methods
```

Comment:
again, see `Exchange#getOrders` for a better chunk streaming pattern.

Reply:
done

Reply:
`getProducts` still queries the count up-front

Reply:
getProducts is done.

Reply:
so getAllLoans should use a different pattern then?

Reply:
anyway, getAllLoans should not even be called for normal users, it is only for admins in the under-the-hood section